evm/executer.go

```diff
@@ -0,0 +1,64 @@
+package evm
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	gethabi "github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/holiman/uint256"
+)
+
+type MonomerContractExecuter struct {
+	ABI          *gethabi.ABI
+	evm          *vm.EVM
+	contractAddr common.Address
+}
+
+type CallParams struct {
+	Sender *common.Address
+	Value  *big.Int
+	Data   []byte
+}
+
+func NewMonomerContractExecuter(evm *vm.EVM, abiJSON string, contractAddr common.Address) (*MonomerContractExecuter, error) {
+	abi, err := gethabi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return nil, err
+	}
+	return &MonomerContractExecuter{
+		ABI:          &abi,
+		evm:          evm,
+		contractAddr: contractAddr,
+	}, nil
+}
+
+func (e *MonomerContractExecuter) Call(params *CallParams) ([]byte, error) {
+	sender, value, data := params.getCallParams()
+	res, _, err := e.evm.Call(
+		vm.AccountRef(sender),
+		e.contractAddr,
+		data,
+		e.evm.Context.GasLimit,
+		value,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("call contract %v: %v", e.contractAddr, err)
+	}
+	return res, nil
+}
+
+func (params *CallParams) getCallParams() (common.Address, *uint256.Int, []byte) {
+	if params.Sender == nil {
+		params.Sender = &MonomerEVMTxOriginAddress
```

Comment:
does the sender affect the withdrawal hash?

Reply:
Yeah it will, the [withdrawal hash](https://github.com/ethereum-optimism/optimism/blob/62c7f3b05a70027b30054d4c8974f44000606fb7/packages/contracts-bedrock/contracts/L2/L2ToL1MessagePasser.sol#L103-L112) is the keccak256 hash of the ABI encoded `WithdrawalTransaction` Solidity struct. The `WithdrawalTransaction` struct contains the tx nonce, sender, target, value, gasLimit, and data.

Reply:
Can we set it to the Cosmos address that sent the original withdrawal tx on the Cosmos side? Maybe we should have address adapters too. We already have `EvmToCosmos` I think.

Reply:
I'm not sure if this causes any problems. I can't think of a case where it does. Does it affect who can withdraw on the L1 side? If not, we can add this to the task board for a later increment.

Reply:
I think we should already be doing that for the `InitiateWithdrawal` tx. The sender is [pulled directly](https://github.com/polymerdao/monomer/pull/91/files#diff-d371e466d16c4d92dff29b2806e847c619149cd6de147f1cb9ae7e9511d5da7fR273) off of the cosmos message and then is also used as `msg.sender` [in the monomer EVM](https://github.com/polymerdao/monomer/pull/91/files#diff-8cc76b058dc98fb1dcad70a01907c83f6304526cb4b1388d5cdfa0c5083b7b43R47). The cosmos address (`sdk.AccAddress`) can be ported directly to the eth address (`common.Address`) since they're both represented as `[20]byte`.

Reply:
Ah my bad, you're right. I missed that call somehow. Thanks